### ISSUE TITLE
Mssql pyodbc datetime with timezone fail to convert at UPDATE

### ISSUE THREAD

**Issue by potiuk** [sqlalchemy/sqlalchemy#7660]
### Describe the bug

When setting the `UTCDate` field in MsSQL, using mssql PyODBC driver as of 1.4.10 fails to update datetime with timezone.

NOTE! While the problem is very similar to one described in #6366, (including identical error), it has not been fixed in 1.4.12, when #6366 has been merged. I believe the problem is that the behaviour described there related to INSERT, our problem in Apache Airflow is about UPDATE.

The problem is that when the UTCDateTime field is updated in MSSQL with PyODBC driver, the value gets converted to string and the string fails to be converted to Date.

The field is defined as UTCDateTime: 

https://github.com/apache/airflow/blob/d353f023ff8856c00b9f054526cb2e40ff0116ae/airflow/models/taskinstance.py#L376

We set the field in one of our tests to ```timezone.utcnow() - datetime.timedelta(seconds=60)```

https://github.com/apache/airflow/blob/d353f023ff8856c00b9f054526cb2e40ff0116ae/tests/jobs/test_scheduler_job.py#L3467

The subsequent `session.flush()` generates an error message: 

> Conversion failed when converting date and/or time from character string. (241) (SQLExecDirectW)')

Caused by this query:
```
UPDATE task_instance SET state=?, trigger_timeout=? WHERE task_instance.task_id = ? AND task_instance.dag_id = ? AND task_instance.run_id = ? AND task_instance.map_index = ?'                                                                  
parameters = ((<TaskInstanceState.DEFERRED: 'deferred'>, '2022-02-02 15:04:07.013951 +00:00', 'dummy1', 'test_timeout_triggers', 't...tanceState.DEFERRED: 'deferred'>, '2022-02-02 15:06:07.013964 +00:00', 'dummy1', 'test_timeout_triggers', 'test2', -1))
```

We bisected the problem and the test clearly stops working in  1.4.10 (1.4.9 is fine) and all versions from 1.4.10 until latest released 1.4.31 exhibit the same problem. NOTE - SQLAlchemy 1.3 works fine as well.

You can see an example failure in CI here: https://github.com/apache/airflow/runs/5027548731?check_suite_focus=true

We are going to temporarily limit SQLAlchemy to <1.4.10 to contain the issue (it breaks our PRs currently) but it would be great to fix it soon-ish so that we can use latest and greates SQLAlchmy (MSSQL is not yet offixially released by Apache Airflow but we plan to do it in ~ 2 months and this issue is effectively blocking us from upgrading SQLAlchemy also for MySQL and Postgres as we keep one version of dependencies for our package - regardless of the database used).

### To Reproduce

1) Clone Apache Airflow GitHub project (https://github.com/apache/airflow/) 

2) In the repository, run `./breeze --github-image-id 2652c6ef435e501e171062398874b38c375f9403 --backend mssql --python 3.9 --db-reset --skip-mounting-local-sources --test-type Core  shell` 

This will bring up a complete development environment with Airflow + MSSQL database via docker-compose (you need to have modern docker and docker-compose installed). It might take few minutes (depending on your network speed) to download the images and start `breeze` environment (MacOS/Linux are supported as Host environment in general)

3) Once this is complete, you get into the `Breeze` shell you should be able to run the tests:

```
root@a6c901f19630:/opt/airflow# 
```

4) Check your SQLALchemy version (should be 1.4.31)

```
root@a6c901f19630:/opt/airflow# pip freeze | grep SQLAlchemy
Flask-SQLAlchemy==2.5.1
SQLAlchemy==1.4.31
SQLAlchemy-JSONField==1.0.0
SQLAlchemy-Utils==0.38.2
root@a6c901f19630:/opt/airflow# 
```

5) You can run the test that failed:

pytest tests/jobs/test_scheduler_job.py::TestSchedulerJob::test_timeout_triggers

It should throw an error:
```
ERROR tests/jobs/test_scheduler_job.py::TestSchedulerJob::test_timeout_triggers - sqlalchemy.exc.PendingRollbackError: This Session's transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original except...
FAILED tests/jobs/test_scheduler_job.py::TestSchedulerJob::test_timeout_triggers - sqlalchemy.exc.DataError: (pyodbc.DataError) ('22007', '[22007] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Conversion failed when converting date and/or time from character string. (241) (SQLExecDirectW)')
=================================================================================================================================== 1 failed, 40 warnings, 1 error in 12.67s ====================================================================================================================================
```

6) Update to SQLALchemy 1.4.9
```
pip install SQLAlchemy==1.4.9
```

7) Re-run the test:

```
pytest tests/jobs/test_scheduler_job.py::TestSchedulerJob::test_timeout_triggers
```

It should succed this time

```
====== 1 passed, 34 warnings in 12.00s =====
```

You can install other versions (including 1.4.12 which was supposed to fix #6366 which seems very closely related). Both #6366 and this one were introduced in 1.4.10.

### Error

```
  tests/conftest.py:630: 
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  tests/conftest.py:595: in cleanup
      for attempt in run_with_db_retries(logger=self.log):
  /usr/local/lib/python3.9/site-packages/tenacity/__init__.py:382: in __iter__
      do = self.iter(retry_state=retry_state)
  /usr/local/lib/python3.9/site-packages/tenacity/__init__.py:349: in iter
      return fut.result()
  /usr/local/lib/python3.9/concurrent/futures/_base.py:439: in result
      return self.__get_result()
  /usr/local/lib/python3.9/concurrent/futures/_base.py:391: in __get_result
      raise self._exception
  tests/conftest.py:603: in cleanup
      self.session.query(SerializedDagModel).filter(
  /usr/local/lib/python3.9/site-packages/sqlalchemy/orm/query.py:3191: in delete
      result = self.session.execute(
  /usr/local/lib/python3.9/site-packages/sqlalchemy/orm/session.py:1691: in execute
      conn = self._connection_for_bind(bind)
  /usr/local/lib/python3.9/site-packages/sqlalchemy/orm/session.py:1532: in _connection_for_bind
      return self._transaction._connection_for_bind(
  /usr/local/lib/python3.9/site-packages/sqlalchemy/orm/session.py:721: in _connection_for_bind
      self._assert_active()
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  
  self = <sqlalchemy.orm.session.SessionTransaction object at 0x7f0b9019a0d0>
  prepared_ok = False, rollback_ok = False, deactive_ok = False
  closed_msg = 'This transaction is closed'
  
      def _assert_active(
          self,
          prepared_ok=False,
          rollback_ok=False,
          deactive_ok=False,
          closed_msg="This transaction is closed",
      ):
          if self._state is COMMITTED:
              raise sa_exc.InvalidRequestError(
                  "This session is in 'committed' state; no further "
                  "SQL can be emitted within this transaction."
              )
          elif self._state is PREPARED:
              if not prepared_ok:
                  raise sa_exc.InvalidRequestError(
                      "This session is in 'prepared' state; no further "
                      "SQL can be emitted within this transaction."
                  )
          elif self._state is DEACTIVE:
              if not deactive_ok and not rollback_ok:
                  if self._rollback_exception:
  >                   raise sa_exc.PendingRollbackError(
                          "This Session's transaction has been rolled back "
                          "due to a previous exception during flush."
                          " To begin a new transaction with this Session, "
                          "first issue Session.rollback()."
                          " Original exception was: %s"
                          % self._rollback_exception,
                          code="7s2a",
  E                       sqlalchemy.exc.PendingRollbackError: This Session's transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original exception was: (pyodbc.DataError) ('22007', '[22007] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Conversion failed when converting date and/or time from character string. (241) (SQLExecDirectW)')
  E                       [SQL: UPDATE task_instance SET state=?, trigger_timeout=? WHERE task_instance.task_id = ? AND task_instance.dag_id = ? AND task_instance.run_id = ? AND task_instance.map_index = ?]
  E                       [parameters: ((<TaskInstanceState.DEFERRED: 'deferred'>, '2022-02-01 20:30:10.018512 +00:00', 'dummy1', 'test_timeout_triggers', 'test', -1), (<TaskInstanceState.DEFERRED: 'deferred'>, '2022-02-01 20:32:10.018530 +00:00', 'dummy1', 'test_timeout_triggers', 'test2', -1))]
  E                       (Background on this error at: https://sqlalche.me/e/14/9h9h) (Background on this error at: https://sqlalche.me/e/14/7s2a)
  
  /usr/local/lib/python3.9/site-packages/sqlalchemy/orm/session.py:601: PendingRollbackError

``` 

More context here: https://github.com/apache/airflow/runs/5027548731?check_suite_focus=true#step:7:6283

### Versions

- OS: Debian Buster (docker image)
- Python: 3.9
- SQLAlchemy: 1.4.10 - 1.4.31 (all versions exhibit the same problem). 1.4.9 is fine.
- Database: MSSQL=2017-latest
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): pyodbc

Name: pyodbc
Version: 4.0.32
Summary: DB API Module for ODBC
Home-page: https://github.com/mkleehammer/pyodbc
Author: 
Author-email: 
License: MIT
Location: /usr/local/lib/python3.9/site-packages

**Comment by zzzeek** [comment]
hi there -

unfortunately we don't have the resources here to install and debug third party application servers like Apache Airflow, we can only fix issues that are shown to be local to SQLAlchemy itself for which we require simple [MCVE](http://stackoverflow.com/help/mcve) scripts to be communicated.  These scripts not only distill the issue down to simple use of SQLAlchemy's API, they also encourage the reporter to distill the actual problem they are having in their application and/or test suite into its fundamentals, often clarifying exactly where the problem actually is (which is sometimes not within SQLAlchemy itself). 

Towards this goal, please provide a succinct, single file .py script with no external dependencies that illustrates the problem you're having.   Feel free to review the [great mcve](https://github.com/sqlalchemy/sqlalchemy/issues?q=is%3Aissue+label%3A%22great+mcve%22+) tag for examples.  thanks!

**Comment by zzzeek** [comment]
IIUC I can take the test at #6366 and make it use an update, is that it?

**Comment by zzzeek** [comment]
Here's that completed into an MCVE, which does not produce an error in the latest 1.4 series.  If you can modify this to show your error, that will be what we need. thanks!

```py
from datetime import datetime
import time

import pytz

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import DateTime
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import Session

Base = declarative_base()


class Project(Base):
    """
    Simple table describing projects.
    """

    __tablename__ = "project"

    id = Column("project_id", Integer, primary_key=True)
    name = Column(String, nullable=False)
    description = Column(String, nullable=False)
    created = Column(DateTime, nullable=False)
    modified = Column(DateTime, nullable=False)


e = create_engine(
    "mssql+pyodbc://scott:tiger^5HHH@mssql2017:1433/test?driver=ODBC+Driver+13+for+SQL+Server",
    echo="debug",
)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)
session = Session(e)
p1 = Project(
        name="name",
        description="desc",
        created=datetime.now(tz=pytz.timezone("utc")),
        modified=datetime.now(tz=pytz.timezone("utc")),
    )
session.add(p1)
session.commit()

time.sleep(1)
p1.modified  = datetime.now(tz=pytz.timezone("utc"))
session.flush()

```

the UPDATE at the end is:

```
UPDATE project SET modified=? WHERE project.project_id = ?
2022-02-02 12:27:17,038 INFO sqlalchemy.engine.Engine [generated in 0.00054s] (datetime.datetime(2022, 2, 2, 17, 27, 17, 18923, tzinfo=<UTC>), 1)
```

**Comment by potiuk** [comment]
Thanks for those. I will try to reproduce it with the script. It **might** be it comes from the way we customize the DateTime field to make it UTC only - I thought initially it's a native SQLAlchemy type, but it's something we did at Airflow. So it might well be it's this (though what I found surprising is that it works fine of Postgres/MySQL/SQLite and only fails with Sqlalchemy 4.1.10+ . I will keep you posted.

**Comment by zzzeek** [comment]
Haven't looked but often these kinds of regressions are caused by custom datatypes overriding the type object directly and implementing bind_processor() and/or result_processor() , when what they should do it to use  [TypeDecorator](https://docs.sqlalchemy.org/en/14/core/custom_types.html?highlight=typedecorator#sqlalchemy.types.TypeDecorator) .  make sure any custom datatype that is augmenting the bound/return value of an existing SQLAlchemy datatype is based on TypeDecorator.

**Comment by potiuk** [comment]
Yeah. We already use TypeDecorator. I am looking into it (I have not written the code myself) but we do have a weird pendulum/datetime duality in Airflow so this might be connected.

Don't want to take more of your time (thaks for the super-quick response BTW!) - but if you could take a look and maybe you will immediately see a problem with our implementation, that would be awesome

https://github.com/apache/airflow/blob/d353f023ff8856c00b9f054526cb2e40ff0116ae/airflow/utils/sqlalchemy.py#L40

**Comment by potiuk** [comment]
BTW. So far I could not reproduce it with the minimal script so I must dig deeper :(

**Comment by potiuk** [comment]
I will come back to it tomorrow. Got some other things to look at (and for now we are limiting SQLALchemy to 1.4.9 to workaround it). But I will come back with more finding.

**Comment by CaselIT** [comment]
please reopen if you can reporduce

**Comment by potiuk** [comment]
For anyone looking at the issue the problem seemed to come from  different treatment of timezone fields  in our custom decorator. the original class is here:

```
class UtcDateTime(TypeDecorator):
    """
    Almost equivalent to :class:`~sqlalchemy.types.DateTime` with
    ``timezone=True`` option, but it differs from that by:
    - Never silently take naive :class:`~datetime.datetime`, instead it
      always raise :exc:`ValueError` unless time zone aware value.
    - :class:`~datetime.datetime` value's :attr:`~datetime.datetime.tzinfo`
      is always converted to UTC.
    - Unlike SQLAlchemy's built-in :class:`~sqlalchemy.types.DateTime`,
      it never return naive :class:`~datetime.datetime`, but time zone
      aware value, even with SQLite or MySQL.
    - Always returns DateTime in UTC
    """
    impl = DateTime(timezone=True)
    cache_ok = True
    def process_bind_param(self, value, dialect):
        if value is not None:
            if not isinstance(value, datetime.datetime):
                raise TypeError('expected datetime.datetime, not ' + repr(value))
            elif value.tzinfo is None:
                raise ValueError('naive datetime is disallowed')
            # For mysql we should store timestamps as naive values
            # Timestamp in MYSQL is not timezone aware. In MySQL 5.6
            # timezone added at the end is ignored but in MySQL 5.7
            # inserting timezone value fails with 'invalid-date'
            # See https://issues.apache.org/jira/browse/AIRFLOW-7001
            if using_mysql:
                from airflow.utils.timezone import make_naive
                return make_naive(value, timezone=utc)
            return value.astimezone(utc)
        return None
    def process_result_value(self, value, dialect):
        """
        Processes DateTimes from the DB making sure it is always
        returning UTC. Not using timezone.convert_to_utc as that
        converts to configured TIMEZONE while the DB might be
        running with some other setting. We assume UTC datetimes
        in the database.
        """
        if value is not None:
            if value.tzinfo is None:
                value = value.replace(tzinfo=utc)
            else:
                value = value.astimezone(utc)

        return value
```

The solution was to change the DateTime implementaiton for mssql to return mssql.DATETIME2(precision=6) instead of the built-in DateTime type.

```
class UtcDateTime(TypeDecorator):
    """
    Almost equivalent to :class:`~sqlalchemy.types.DateTime` with
    ``timezone=True`` option, but it differs from that by:
    - Never silently take naive :class:`~datetime.datetime`, instead it
      always raise :exc:`ValueError` unless time zone aware value.
    - :class:`~datetime.datetime` value's :attr:`~datetime.datetime.tzinfo`
      is always converted to UTC.
    - Unlike SQLAlchemy's built-in :class:`~sqlalchemy.types.DateTime`,
      it never return naive :class:`~datetime.datetime`, but time zone
      aware value, even with SQLite or MySQL.
    - Always returns DateTime in UTC
    """
    impl = DateTime(timezone=True)
    cache_ok = True
    def process_bind_param(self, value, dialect):
        if value is not None:
            if not isinstance(value, datetime.datetime):
                raise TypeError('expected datetime.datetime, not ' + repr(value))
            elif value.tzinfo is None:
                raise ValueError('naive datetime is disallowed')
            # For mysql we should store timestamps as naive values
            # Timestamp in MYSQL is not timezone aware. In MySQL 5.6
            # timezone added at the end is ignored but in MySQL 5.7
            # inserting timezone value fails with 'invalid-date'
            # See https://issues.apache.org/jira/browse/AIRFLOW-7001
            if using_mysql:
                from airflow.utils.timezone import make_naive
                return make_naive(value, timezone=utc)
            return value.astimezone(utc)
        return None
    def process_result_value(self, value, dialect):
        """
        Processes DateTimes from the DB making sure it is always
        returning UTC. Not using timezone.convert_to_utc as that
        converts to configured TIMEZONE while the DB might be
        running with some other setting. We assume UTC datetimes
        in the database.
        """
        if value is not None:
            if value.tzinfo is None:
                value = value.replace(tzinfo=utc)
            else:
                value = value.astimezone(utc)

        return value

    def load_dialect_impl(self, dialect):
        if dialect.name == 'mssql':
            return mssql.DATETIME2(precision=6)
        return super().load_dialect_impl(dialect)
```
